### ISSUE TITLE
Schematic Python Requirements

### ISSUE THREAD

**Issue by GiaJordan** [Sage-Bionetworks/schematic#925]
**Describe the solution you'd like**
Discussion and decision on how many and which versions of python should be supported. Currently, versions `3.7`, `3.8`, `3.9`, and `3.10` are supported. Restricting to the latest python versions as reasonable can reduce issues encountered during installation and ensure we are receiving the benefits of using more up-to-date features included in newer python versions.

**How important is this feature?** Select from the options below:
• 🌗 Medium - can do work without it; but it's important (e.g. to save time or for convenience)

**When will use cases depending on this become relevant?** Select from the options below:
• Mid-term - 2-4 months

**Additional context**

**Comment by GiaJordan** [comment]
Bumping so we can prioritize ahead of python `v3.7` [end of life](https://devguide.python.org/versions/) on June 27 of this year
@milen-sage @MiekoHash

**Comment by MiekoHash** [comment]
Like! Thanks, @GiaJordan.

**Comment by milen-sage** [comment]
That's a good point! Let's put this on our horizon for triage in sprint 2. We can evaluate if we are ready to implement or can wait.

**Comment by MiekoHash** [comment]
@milen-sage I moved this to sprint 2, assigned prio per original assessment, and left it unassigned.

**Comment by milen-sage** [comment]
Thank you @MiekoHash - we will triage the issue in sprint 2.

**Comment by linglp** [comment]
If we want to sync schematic develop branch with the AWS deployment, and if we switch to Python 11.0, we have to find a secure python image (without high/critical level of security risks). But currently the image [here](https://snyk.io/advisor/docker/python/3.11) seems to be not secure. If we definitely want to switch, I will have to look into a secure python docker image.

**Comment by milen-sage** [comment]
From @andrewelamb schematic_db requires Python 3.9.

**Comment by GiaJordan** [comment]
For this issue, I'll focus on depreciating versions 3.7 and 3.8. I created a new issue on updating to 3.11 (#1091) that we can complete separately as it's not needed to unblock table upserts.
@milen-sage @MiekoHash @linglp

**Comment by mialy-defelice** [comment]
Tagging for visibility:

@milen-sage @linglp @GiaJordan @andrewelamb 

Based on our discussion yesterday here are my notes on having Python issues. I will try to organize my thoughts as much as possible but let me know if something is unclear.

TL;DR Scroll to the bottom for the solution :)

## The Problem: 
### `poetry install` does not work

- The crux of the problem is, there were updates to the poetry lock file and I needed to do a new `poetry install`.
- My version of poetry was old, I needed `1.2.0` so I was unable to do the install like expected.
- I was unable to do a clean update or uninstall because all of the older install/uninstall method was depreciated and the new uninstall method will not work with the old install. This is related to[ this GitHub issue](https://sagebionetworks.slack.com/archives/C01ANC02U59/p1664369561664879). 
- Since I could not use any standard uninstallers I attempted to manually uninstall poetry but was running into issues. 
- To overcome the above uninstall issue I decided to delete as much as I could then `pipx install poetry` since others had success using `pipx`.
- This all leads to the next issue.

### `pipx install poetry` uses the newest version of python to install poetry.

- Without specifying to do this, `pipx` will automatically use the newest version of python to install poetry and peg your poetry to this version, even if you specify a different python in your `.toml` file. 
- If you then try to install your poetry package you will get an error like this bc we are not allowing the newest version of python:
```mialydefelice@Mialys-MBP schematic % poetry install
The currently activated Python version 3.11.1 is not supported by the project (>=3.7.1,<3.11).
Trying to find and use a compatible version. 
Using python3 (3.8.2)
Installing dependencies from lock file

The current project's Python requirement (3.11.1) is not compatible with some of the required packages Python requirement:
  - numpy requires Python >=3.7,<3.11, so it will not be satisfied for Python 3.11.1

Because numpy (1.21.6) requires Python >=3.7,<3.11
 and no versions of numpy match >=1.21.1,<1.21.6 || >1.21.6,<2.0.0, numpy is forbidden.
So, because schematicpy depends on numpy (^1.21.1), version solving failed.

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
```

- I then decided to uninstall poetry installed through pipx `pipx uninstall poetry` (this is very easy) and installed through poetry curl installation `curl -sSL https://install.python-poetry.org | python3 -` but this also failed!

```mialydefelice@Mialys-MBP schematic % curl -sSL https://install.python-poetry.org/ | python3 -
Retrieving Poetry metadata

# Welcome to Poetry!

This will download and install the latest version of Poetry,
a dependency and package manager for Python.

It will add the `poetry` command to Poetry's bin directory, located at:

/Users/mialydefelice/.local/bin

You can uninstall at any time by executing this script with the --uninstall option,
and these changes will be reverted.

Installing Poetry (1.3.2): An error occurred. Removing partial environment.
Poetry installation failed.
See /Users/mialydefelice/Documents/schematic_b/schematic/poetry-installer-error-_u87dcug.log for error logs.
```
The logs:
```
dyld: Library not loaded: @executable_path/../Python3
  Referenced from: /Users/mialydefelice/Library/Application Support/pypoetry/venv/bin/python3
  Reason: image not found

Traceback:

  File "<stdin>", line 919, in main
  File "<stdin>", line 552, in run
```
## Attempts at finding a solution:
I will not go through every single attempt I made, but will just highlight why I think `pipx` is a problem child.
- For some reason after my last error trying to use the standard poetry install, I decided that since others were using `pipx` I would try to debug that install. It seemed simple enough, its using the wrong python, ill just fix that python issue. 
- Moral of the story...this is essentially an impossible task. 
- After I finally got the correct python to be used: 
  - Delete all poetry cache and artifact files manually (`pyenv clear cache --all` does not work to clear the artifacts)
  - `pyenv shell 3.10.9`
  - `poetry env use 3.10.9` 
  - `poetry install`
- there was no path created to save the eventual `schematicpy` file and the `poetry install` would fail in the end with the following error.
```
Installing the current project: schematicpy (1.0.0)  - Failed to create schematicpy.pth for /Users/mialydefelice/Documents/schematic_c/schematic

Unable to access any of "/Users/mialydefelice/.local/pipx/venvs/poetry/lib/python3.10/site-packages/schematicpy-1.0.0.dist-info"
```
This is because even if you do everything right to get to the right python version a path to python3.10 site packages (`/Users/mialydefelice/.local/pipx/venvs/poetry/lib/python3.10`) is never even created... I also don't know how to get this to happen.

- After many many many attempts at a solution I realized my only recourse was to go _back_ to the standard poetry install using curl.

## The Solution:
- If you have previously installed poetry:
  - Uninstall poetry: `pipx uninstall poetry`
  - Delete all cache and artifact files.
- Install pyenv, set your local python and _then_ install poetry 
  - Download pyenv (on Mac can do this via Homebrew):
    - `brew install pyenv`
  - `pyenv install 3.10.9` (or whatever version of python is needed)
  - `pyenv local 3.10.9`
  - `curl -sSL https://install.python-poetry.org | python3 -`
  - `poetry debug info` (to check consistent python install versions)
  - cd to `schematic`
  - `poetry install`
